### PR TITLE
SOLR-11904: Mark ReplicationHandler's polling thread as a Solr server thread so the PKI Interceptor is activated to   allow PULL replicas to replicate from security-enabled leaders

### PR DESCRIPTION
**Pull request by thelabdude**
# Description

Fixes SOLR-11904

# Solution

The PKI interceptor relies on a check that the requesting thread is a Solr server thread. The ReplicationHandler's polling thread is not created using the ExecutorUtil (since it's a scheduled), so needs to explicitly mark itself as a server thread so the PKI interceptor kicks in for calling out to leaders.

# Tests

Added unit test `TestPullReplicaWithAuth`; unfortunately, I'm leaving `TestPullReplica` disabled with `AwaitsFix` b/c it was still failing occasionally with beasting locally :-(

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Review by dsmiley — APPROVED** [review]

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java [hunk 1]
```diff
@@ -97,6 +92,90 @@ public static void tearDownCluster() {
     TestInjection.reset();
   }
 
+  static void waitForDeletion(String collection) throws InterruptedException, KeeperException {
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java:95]: nit: did this need to move, or can it stay around line 555 to minimize the diff? All of these methods moved around but it looks like they mostly stayed the same. Feel free to move them in a follow on PR :)

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java [hunk 2]
```diff
@@ -122,15 +201,15 @@ public void tearDown() throws Exception {
     super.tearDown();
   }
 
-  @Repeat(iterations=2) // 2 times to make sure cleanup is complete and we can create the same collection
-  // commented out on: 17-Feb-2019   @BadApple(bugUrl="https://issues.apache.org/jira/browse/SOLR-12028") // 21-May-2018
+  @Repeat(iterations = 2) // 2 times to make sure cleanup is complete and we can create the same collection
   public void testCreateDelete() throws Exception {
+    final int rand = 0; // random().nextInt(3);
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java:206]: Dubious

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java [hunk 3]
```diff
@@ -191,91 +270,56 @@ public void testCreateDelete() throws Exception {
...
-        }
-      }
-    }
-  }
-
   @SuppressWarnings("unchecked")
-  // 12-Jun-2018 @BadApple(bugUrl="https://issues.apache.org/jira/browse/SOLR-12028")
   public void testAddDocs() throws Exception {
-    int numPullReplicas = 1 + random().nextInt(3);
+    int numPullReplicas = 1; // + random().nextInt(3);
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java:275]: Does this fail with more than 1?

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java [hunk 4]
```diff
@@ -191,91 +270,56 @@ public void testCreateDelete() throws Exception {
...
+    CloudSolrClient solrClient = cluster.getSolrClient();
+    for (int i = 0; i < 5; i++) {
       numDocs++;
-      cluster.getSolrClient().add(collectionName, new SolrInputDocument("id", String.valueOf(numDocs), "foo", "bar"));
-      cluster.getSolrClient().commit(collectionName);
+
+      UpdateRequest ureq = new UpdateRequest();
+      ureq.add(new SolrInputDocument("id", String.valueOf(numDocs), "foo", "bar"));
+      ureq.process(solrClient, collectionName);
+      ureq.commit(solrClient, collectionName);
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java:291]: just curious why this needed to be split out into two calls? if it's just for debugging, or there's some condition under test. leave a comment either way.

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java [hunk 5]
```diff
@@ -191,91 +270,56 @@ public void testCreateDelete() throws Exception {
...
-              "qt", "/admin/plugins",
-              "stats", "true");
+          SolrQuery req = new SolrQuery("qt", "/admin/plugins", "stats", "true");
           QueryResponse statsResponse = pullReplicaClient.query(req);
+          // adds is a gauge, which is null for PULL replicas
+          assertNull("Replicas shouldn't process the add document request: " + statsResponse,
+              ((Map<String, Object>) (statsResponse.getResponse()).findRecursive("plugins", "UPDATE", "updateHandler", "stats")).get("UPDATE.updateHandler.adds"));
           assertEquals("Replicas shouldn't process the add document request: " + statsResponse,
-              0L, ((Map<String, Object>)(statsResponse.getResponse()).findRecursive("plugins", "UPDATE", "updateHandler", "stats")).get("UPDATE.updateHandler.adds"));
+              0L, ((Map<String, Object>) (statsResponse.getResponse()).findRecursive("plugins", "UPDATE", "updateHandler", "stats")).get("UPDATE.updateHandler.cumulativeAdds.count"));
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java:309]: Can we pull out the common parts of these lines?

**Comment by thelabdude** [comment]
Going to watch the new unit test in nightlies before back-porting to 8.x

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplicaWithAuth.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+@LogLevel("org.apache.solr.handler.ReplicationHandler=DEBUG,org.apache.solr.handler.IndexFetcher=DEBUG")
+public class TestPullReplicaWithAuth extends SolrCloudTestCase {
+
+  private static final String USER = "solr";
+  private static final String PASS = "SolrRocksAgain";
+  private static final String collectionName = "testPullReplicaWithAuth";
+
+  @BeforeClass
+  public static void setupCluster() throws Exception {
+    final String SECURITY_JSON = Utils.toJSONString
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/cloud/TestPullReplicaWithAuth.java:65]: Can you try this with Map.of, which is more type safe? I've got similar stuff going on in #107

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplicaWithAuth.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+    configureCluster(3)
+        .addConfig("conf", configset("cloud-minimal"))
+        .withSecurityJson(SECURITY_JSON)
+        .configure();
+  }
+
+  @AfterClass
+  public static void tearDownCluster() {
+    TestInjection.reset();
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/cloud/TestPullReplicaWithAuth.java:83]: don't think we need this, I think it's automatically done for us

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplicaWithAuth.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  @SuppressWarnings({"rawtypes"})
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/cloud/TestPullReplicaWithAuth.java:86]: please don't do this.

File: solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java (hunk 3 above)
Comment (thelabdude) [reply solr/core/src/test/org/apache/solr/cloud/TestPullReplica.java:275]: test is flaky, so the whole test class is still disabled ... I tried to figure out what was causing the failures but couldn't nail it down

**Pull request merged**